concrete-core/src/crypto/gsw/ciphertext.rs

```diff
@@ -0,0 +1,591 @@
+    ///             lwe.as_mut_tensor().fill_with_element(9);
+    ///         }
+    ///     });
+    /// assert!(gsw.as_tensor().iter().all(|a| *a == 9));
+    /// assert_eq!(gsw.level_matrix_iter_mut().count(), 3);
+    /// ```
+    #[cfg(feature = "multithread")]
+    pub fn par_level_matrix_iter_mut(
+        &mut self,
+    ) -> impl IndexedParallelIterator<Item = GswLevelMatrix<&mut [<Self as AsRefTensor>::Element]>>
+    where
+        Self: AsMutTensor,
+        <Self as AsMutTensor>::Element: Sync + Send,
+    {
+        let chunks_size = self.lwe_size.0 * self.lwe_size.0;
+        let lwe_size = self.lwe_size;
+        self.as_mut_tensor()
+            .par_subtensor_iter_mut(chunks_size)
+            .enumerate()
+            .map(move |(index, tensor)| {
+                GswLevelMatrix::from_container(
+                    tensor.into_container(),
+                    lwe_size,
+                    DecompositionLevel(index + 1),
+                )
+            })
+    }
+
+    /// Computes the external product and adds it to the output
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use concrete_commons::dispersion::LogStandardDev;
+    ///
+    /// use concrete_commons::parameters::LweDimension;
+    /// use concrete_commons::parameters::LweSize;
+    /// use concrete_core::crypto::encoding::Plaintext;
+    /// use concrete_core::crypto::gsw::GswCiphertext;
+    /// use concrete_core::crypto::lwe::LweCiphertext;
+    /// use concrete_core::crypto::secret::LweSecretKey;
+    /// use concrete_core::crypto::secret::generators::{
+    ///     EncryptionRandomGenerator, SecretRandomGenerator,
+    /// };
+    /// use concrete_commons::parameters::DecompositionLevelCount;
+    /// use concrete_commons::parameters::DecompositionBaseLog;
+    ///
+    /// let mut secret_generator = SecretRandomGenerator::new(None);
+    /// let mut encryption_generator = EncryptionRandomGenerator::new(None);
+    ///
+    /// let lwe_sk = LweSecretKey::generate_binary(LweDimension(256), &mut secret_generator);
+
+    /// let mut gsw = GswCiphertext::allocate(
+    ///     0 as u32,
+    ///     LweSize(257),
+    ///     DecompositionLevelCount(3),
+    ///     DecompositionBaseLog(7),
+    /// );
+    /// let std_dev = LogStandardDev(-20.);
+    /// lwe_sk.encrypt_constant_gsw(
+    ///     &mut gsw,
+    ///     &Plaintext(1 as u32),
+    ///     std_dev,
+    ///     &mut encryption_generator,
+    /// );
+    ///
+    /// let mut ciphertext = LweCiphertext::allocate(0 as u32, LweSize(257));
+    /// let mut res = LweCiphertext::allocate(0 as u32, LweSize(257));
+    ///
+    /// lwe_sk.encrypt_lwe(
+    ///     &mut ciphertext,
+    ///     &Plaintext(0 as u32),
+    ///     std_dev,
+    ///     &mut encryption_generator,
+    /// );
+    ///
+    /// gsw.external_product(&mut res, &ciphertext);
+    /// ```
+    pub fn external_product<C1, C2>(&self, output: &mut LweCiphertext<C1>, lwe: &LweCiphertext<C2>)
```

Comment:
Ok now I understand why you have a rounded_buffer in the struct. It is true that as of today, we haven't had a clear schema on how to properly integrate those operations which necessitates buffers. I have applied the same schema as yours in the bootstrapping keys, but in the end it feels a bit off. 

The current status of our reflexions on that is that it would be better to have a `GswLweProductEngine` object which would have a `product` method. This would also prevent binding a specific product algorithm to the type itself.

All in all, I don't think now is the right time to settle on this matter, and think we can integrate it that the way it is.